packages/compartment-mapper/src/logical-path.js

```diff
@@ -0,0 +1,343 @@
+/**
+ * Provides {@link LogicalPathGraph} and {@link makeShortestPath}.
+ *
+ * Portions adapted from
+ * {@link https://github.com/datavis-tech/graph-data-structure graph-data-structure},
+ * which is Copyright (c) 2016 Curran Kelleher and licensed under the MIT
+ * License.
+ *
+ * @module
+ */
+
+/**
+ * @import {TraversalContext} from './types/logical-path.js';
+ */
+
+const { stringify: q } = JSON;
+
+/**
+ * Remove the node with the minimum weight from the priority queue.
+ *
+ * Performs linear search.
+ * @template [T=string]
+ * @param {TraversalContext<T>} tracks
+ * @returns {T|undefined}
+ */
+const extractMin = ({ distances, queue }) => {
+  let min = Infinity;
+  let minNode;
+
+  queue.forEach(node => {
+    const nodeWeight = distances.get(node) ?? Infinity;
+
+    if (nodeWeight < min) {
+      min = nodeWeight;
+      minNode = node;
+    }
+  });
+
+  if (minNode === undefined) {
+    queue.clear();
+    return undefined;
+  }
+
+  queue.delete(minNode);
+  return minNode;
+};
+
+/**
+ * @template [T=string]
+ * @param {LogicalPathGraph<T>} graph
+ * @param {TraversalContext<NoInfer<T>>} context
+ * @param {T} source
+ * @param {T} target
+ */
+const relax = (graph, { distances, predecessors }, source, target) => {
+  const number = graph.getEdgeWeight(source, target);
+
+  const distanceSource = distances.get(source);
+  const distanceTarget = distances.get(target);
+
+  assert(
+    distanceSource !== undefined,
+    `Missing distance for source ${q(source)}`,
+  );
+  assert(
+    distanceTarget !== undefined,
+    `Missing distance for target ${q(target)} target`,
+  );
+
+  if (distanceTarget > distanceSource + number) {
+    distances.set(target, distanceSource + number);
+    predecessors.set(target, source);
+  }
+};
+
+/**
+ * Assembles the shortest path by traversing the
+ * predecessor subgraph from destination to source.
+ *
+ * @template [T=string]
+ * @param {TraversalContext<NoInfer<T>>} context Traversal context object
+ * @param {NoInfer<T>} source Source node
+ * @param {NoInfer<T>} target Destination node
+ * @returns {[T, T, ...T[]]} Nodes from `source` to `target` inclusive
+ * @throws If no path is found
+ * @throws If the path has less than two nodes
+ */
+const getPath = ({ predecessors }, source, target) => {
+  /** @type {T[]} */
+  const nodeList = [];
+
+  let node = target;
+
+  while (predecessors.has(node)) {
+    const currentNode = /** @type {T} */ (predecessors.get(node));
+    nodeList.unshift(node);
+    node = currentNode;
+  }
+
+  assert.equal(node, source, `No path found from ${q(source)} to ${q(target)}`);
+
+  nodeList.unshift(node);
+
+  assert(
+    nodeList.length >= 2,
+    `The path from ${source} to ${target} should have a least two nodes`,
+  );
+
+  return /** @type {[T, T, ...T[]]} */ (nodeList);
+};
+
+/**
+ * @template [T=string] The type of nodes in the graph
```

Comment:
The generic is not strictly necessary because we do not use anything but the default. It enables reuse of this data structure for other purposes, though. 🤷